packages/query/src/block-processor.ts

```diff
@@ -103,89 +115,126 @@ export class BlockProcessor implements BlockProcessorInterface {
       // - decrypts new swaps
       // - updates idb with advice
       //const shouldStoreProgress = await this.viewServer.scanBlock(compactBlock);
-      const wasmWantsFlush = await this.viewServer.scanBlock(compactBlock);
+      const scannerWantsFlush = await this.viewServer.scanBlock(compactBlock);
 
       // flushing is slow, avoid it until
       // - wasm says
       // - every 1000th block
       // - every block at tip
       const flushReasons = {
-        wasmWantsFlush,
+        scannerWantsFlush,
         interval: compactBlock.height % 1000n === 0n,
         new: compactBlock.height > latestBlockHeight,
       };
 
       if (flushReasons.new) console.log('finished sync', performance.measure('start'));
 
-      let compactScan;
+      // state commitments identifying new records in this compact block, with a
+      // boolean to indicate if we should recover their source id
+      const blockCommitments = new Map<CommitmentHexString, NeedsSourceId>();
+      let flush;
       if (Object.values(flushReasons).some(Boolean)) {
-        compactScan = this.viewServer.flushUpdates();
-        console.log('flush', compactBlock, flushReasons, compactScan);
+        flush = this.viewServer.flushUpdates();
+        console.log('flush', compactBlock, flushReasons, flush);
 
         // in an atomic query, this
         // - saves 'sctUpdates'
         // - saves new decrypted notes
         // - saves new decrypted swaps
         // - updates last block synced
-        await this.indexedDb.saveScanResult(compactScan);
+        await this.indexedDb.saveScanResult(flush);
 
         // - detect unknown asset types, query the chain for asset metadata
         // - or, generate default fallback metadata
         // - update idb
         // this makes a network query.
         // TODO: should this be scanning swaps as well?
-        await this.identifyNewAssets(compactScan.newNotes);
+        await this.identifyNewAssets(flush.newNotes);
+
+        // boolean indicates if we should recover source id
+        for (const { noteCommitment, source } of flush.newNotes)
+          blockCommitments.set(
+            uint8ArrayToHex(noteCommitment!.inner),
+            source!.equals(blankTxSource),
+          );
+        for (const { swapCommitment, source } of flush.newSwaps)
+          blockCommitments.set(
+            uint8ArrayToHex(swapCommitment!.inner),
+            source!.equals(blankTxSource),
+          );
       }
 
       // nullifiers on this block may match notes or swaps from db
-      // mark them as spent/claimed, return what was used
-      const spentNullifiers = await this.resolveNullifiers(
+      // - update idb, mark as spent/claimed
+      // - return records' commitments
+      // - with a boolean to indicate if we should recover source id
```

Comment:
as far as i can tell: records retrieved this way always contain sources already

Reply:
i believe these commitments are also never present on the current block so i will stop collecting them